packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/ExoPlayerEventListener.java

```diff
@@ -0,0 +1,101 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.plugins.videoplayer;
+
+import androidx.annotation.NonNull;
+import androidx.media3.common.PlaybackException;
+import androidx.media3.common.Player;
+import androidx.media3.common.VideoSize;
+import androidx.media3.exoplayer.ExoPlayer;
+
+final class ExoPlayerEventListener implements Player.Listener {
+  private final ExoPlayer exoPlayer;
+  private final VideoPlayerCallbacks events;
+  private boolean isBuffering = false;
+  private boolean isInitialized = false;
+
+  ExoPlayerEventListener(ExoPlayer exoPlayer, VideoPlayerCallbacks events) {
+    this.exoPlayer = exoPlayer;
+    this.events = events;
+  }
+
+  private void setBuffering(boolean buffering) {
+    if (isBuffering == buffering) {
+      return;
+    }
+    isBuffering = buffering;
+    if (buffering) {
+      events.onBufferingStart();
+      events.onBufferingUpdate(exoPlayer.getBufferedPosition());
+    } else {
+      events.onBufferingEnd();
+    }
+  }
+
+  @SuppressWarnings("SuspiciousNameCombination")
+  private void sendInitialized() {
+    if (isInitialized) {
+      return;
+    }
+    isInitialized = true;
+    VideoSize videoSize = exoPlayer.getVideoSize();
+    int rotationCorrection = 0;
+    int width = videoSize.width;
+    int height = videoSize.height;
+    if (width != 0 && height != 0) {
+      int rotationDegrees = videoSize.unappliedRotationDegrees;
+      // Switch the width/height if video was taken in portrait mode
+      if (rotationDegrees == 90 || rotationDegrees == 270) {
+        width = videoSize.height;
+        height = videoSize.width;
+      }
+      // Rotating the video with ExoPlayer does not seem to be possible with a Surface,
+      // so inform the Flutter code that the widget needs to be rotated to prevent
+      // upside-down playback for videos with rotationDegrees of 180 (other orientations work
+      // correctly without correction).
+      if (rotationDegrees == 180) {
+        rotationCorrection = rotationDegrees;
+      }
+    }
+    events.onInitialized(width, height, exoPlayer.getDuration(), rotationCorrection);
+  }
+
+  @Override
+  public void onPlaybackStateChanged(final int playbackState) {
+    switch (playbackState) {
+      case Player.STATE_BUFFERING:
+        setBuffering(true);
```

Comment:
This is a slight behavior change, right?

Before we would always invoke `videoPlayerEvents.onBufferingUpdate(exoPlayer.getBufferedPosition())`:
```
setBuffering(true);
sendBufferingUpdate();
```
but now we only invoke it if we pass the check at the start of the slightly modified `setBuffering`.

I don't know enough about `video_player` to know if that case (we are in the case of this switch `Player.STATE_BUFFERING` while `isBuffering` is already true) can happen/matters, so feel free to ignore if it isn't important.

Reply:
Not only was this a good catch, but you caught a bug missed in the previous refactor.

I've fixed this and added a test to catch regressions.